apps/consulting/posts/improvements-to-bsr-matrix-multiplication-in-pytorch.md

```diff
@@ -0,0 +1,292 @@
+    name="fig-5"
+    alt="A plot displaying sparsity ratio (as a percentage) on the horizontal axis, and speedup over
+    the dense baseline on the vertical.  There is a horizontal line marking the speedup ratio of 1.0
+    above which the sparse subject takes less time to execute than the dense counterpart.  Three lines are
+    plotted indicating block sizes used for the sparse tensor of 16, 32, and 64."
+/>
+
+Now we look at the performance for half precision data types. The new kernel provides between 20-50x
+speedup over the [older sparse implementation][#fig-6]. While a speedup of this size is always a notable
+result, we have known that the implementation supporting half precision data types would not be
+optimal. However, as cuSPARSE did not support these data types a working implementation was created
+by composing calls into the dense kernel for each sparse block. This approach suffers from a
+significant penalty due to launching many CUDA kernels. 
+
+<img src="posts/improvements-to-bsr-matrix-multiplication-in-pytorch/bsr_dense_mm_half_vs_sparse.png"
+    name="fig-6"
+    alt="A plot displaying sparsity ratio (as a percentage) on the horizontal axis, and speedup over
+    the older sparse implementation on the vertical.  Three lines are plotted indicating block sizes
+    used for the sparse tensor of 16, 32, and 64."
+/>
+
+Comparing performance for half precision to the [dense baseline][#fig-7] we again see one clear win.
+Using a block size of 64 the triton implementation begins to out perform the dense matrix multiply
+at 60% sparsity.  Unfortunately, this does not hold for smaller block sizes with 32 reach >1x
+speedup at 90% sparsity, and block size 16 only passing this point at 99%. 
+
+<img src="posts/improvements-to-bsr-matrix-multiplication-in-pytorch/bsr_dense_mm_half_vs_dense.png"
+    name="fig-7"
+    alt="A plot displaying sparsity ratio (as a percentage) on the horizontal axis, and speedup over
+    the older sparse implementation on the vertical.  There is a horizontal line marking the speedup
+    ratio of 1.0 above which the sparse subject takes less time to execute than the dense
+    counterpart.  Three lines are plotted indicating block sizes used for the sparse tensor of 16,
+    32, and 64."
+/>
+
+## Try it out!
+
+These kernels will be shipped with PyTorch 2.1. They can all be found in the
+`torch.sparse._triton_ops` private submodule. Although not yet stable, the curious user may want
+experiment with these tools. If you want to stick to the public features only, the `bsr_dense_mm`
+kernel is also fully integrated with matrix multiply ops like `torch.mm` and  `torch.addmm`. 
+If you meet the following conditions:
+
+- The RHS of the matrix multiply or `A` in `A@B`, `torch.mm(A, B)`, or `torch.addmm(C, A, B)` has
+  BSR layout, and other tensors are dense.
+- The device type is `cuda`
+- The data type is `torch.float16`, or `torch.bfloat16`
```

Comment:
Not for float32?

Reply:
We wanted to more thoroughly evaluate this kernel before enabling dispatch for `float32` which was available for a long time. The half precision data types were handled by some very poorly performing code that was new in PyTorch 2.0 so it was seen as a lower risk to route through here for those dtypes at first.